pkg/sql/alter_table.go

```diff
@@ -1132,3 +1148,63 @@ func (p *planner) updateFKBackReferenceName(
 	}
 	return errors.Errorf("missing backreference for foreign key %s", ref.Name)
 }
+
+// alterTableOwner sets the owner of the table to newOwner and returns true if the descriptor
+// was updated.
+func (p *planner) alterTableOwner(
+	ctx context.Context, n *alterTableNode, newOwner string,
+) (bool, error) {
+	desc := n.tableDesc
+	privs := desc.GetPrivileges()
+
+	// If the owner we want to set to is the current owner, do a no-op.
+	if newOwner == privs.Owner {
+		return false, nil
+	}
+
+	// Make sure the newOwner exists.
+	roleExists, err := p.RoleExists(ctx, newOwner)
+	if err != nil {
+		return false, err
+	}
+	if !roleExists {
+		return false, pgerror.Newf(pgcode.UndefinedObject, "role/user %s does not exist", newOwner)
+	}
+
+	// The user explicitly has to be owner to change the owner.
+	if ok := IsOwner(desc, p.User()); !ok {
+		return false, pgerror.Newf(pgcode.InsufficientPrivilege,
+			"must be owner of table %s", tree.Name(desc.GetName()))
+	}
+
+	// Requirements from PG:
+	// To alter the owner, you must also be a direct or indirect member of the new owning role,
+	// and that role must have CREATE privilege on the type's schema.
+	memberOf, err := p.MemberOfWithAdminOption(ctx, privs.Owner)
+	if err != nil {
+		return false, err
+	}
+	if _, ok := memberOf[newOwner]; !ok {
+		return false, pgerror.Newf(
+			pgcode.InsufficientPrivilege, "must be member of role %s", newOwner)
+	}
+
+	// Ensure the new owner has CREATE privilege on the type's schema.
+	// We do not have to check for the public schema.
+	if desc.GetParentSchemaID() != keys.RootNamespaceID && desc.GetParentSchemaID() != keys.PublicSchemaID {
```

Comment:
Should I add that now or wait until your PR is merged?
I wouldn't mind waiting for that PR since a lot of that work will be helpful for ALTER SCHEMA OWNER as well, can do the refactor there.

Reply:
Updated.